### ISSUE TITLE
Small launchers/low-capacity missile racks

### ISSUE THREAD

**Issue by Pointedstick** [endless-sky/endless-sky#3372]
Right now fighter-sized craft generally mount a small-ish missile launcher of some type. But really big missiles like Heavy Rockets and Torpedoes require really big launchers, making those types of weapons unavailable to small ships and especially fighters, which historically have existed primarily as delivery platforms for large bombs and missiles in the first place.

I'd like to propose very small gun hardpoint launchers that can only hold 3 or 4 missiles, like today's missile pylons:

![smart-launcher](https://user-images.githubusercontent.com/1097249/33673827-195578e4-da6b-11e7-940b-72c4d0ddf88f.jpg)

![a pylon loaded with agm-86b air launched cruise missiles aclms is positioned under the wing of a 379th bombardment wing b-52 stratofortress aircraft](https://user-images.githubusercontent.com/1097249/48793693-04e2ac80-ecb5-11e8-806b-2a3b5ab2cbf2.jpg)

![jdam_gbu30_mer](https://user-images.githubusercontent.com/1097249/48793775-496e4800-ecb5-11e8-948e-570ce8d91fe2.jpg)

This would dovetail especially well with https://github.com/endless-sky/endless-sky/pull/3350, which increases the number of gun hardpoints available to fighter-sized craft: you could give your fighters two blasters or lasers, one Meteor Missile launcher, and the capacity to fire 3 or 4 Torpedoes or Heavy Rockets via a small pylon launcher.

**Comment by jafdy** [comment]
Do you mean #3350 instead of 335?

**Comment by solardawning** [comment]
Could be done in an interesting manner by making new fighter/drone anti-capital torpedoes that are used in this manner, and allowing them to automatically reload on docking with a mothership that has spares stored in an ammo container. Imagine an auto docking and launch behavior that occurs if all their hardpoints were one shot ammunition, and have been used.

**Comment by tehhowch** [comment]
Auto-docking for reload of conventional ammo is handled in #1971 (which needs a little attention), but for outfits which fire themselves the logic is not so simple (currently). Once the weapon is fired, the ship has no way of knowing it had that outfit, or that it could use that outfit, since the reference to the outfit vanished with it.
https://github.com/endless-sky/endless-sky/blob/224e62f563a55fc57e20b2e851a945836378bd0f/source/Ship.cpp#L2542-L2547
https://github.com/endless-sky/endless-sky/blob/224e62f563a55fc57e20b2e851a945836378bd0f/source/Ship.cpp#L2474-L2497

**Comment by Pointedstick** [comment]
Yes, I was just thinking about that, in fact! I might put together a PR later this week to implement it.

**Comment by Pointedstick** [comment]
What do people think the name should be? E.g., for the Heavy Rocket:
- Heavy Rocket (Hardpoint)
- Hardpoint Heavy Rocket
- Heavy Rocket Pylon
- Something else?

**Comment by Pointedstick** [comment]
Playing with a prototype on my fighters, it's a real pain in the butt to restock, because the outfits count as launchers, not ammo, so the game doesn't prompt you to reload when visiting the outfitter.

I'm trying an alternative approach of making small launcher outfits ("Rocket Pylon") that can hold 2 or 3 missiles apiece. This works around that problem, but another one rears its head: it's possible to buy ammo storage outfits--which are quite small--to boost capacity back up to levels that are unrealistic and overpowered for fighter-sized craft.

**Comment by tehhowch** [comment]
Some method of tracking i-am-my-ammunition outfits, to facilitate the automated restocking, would certainly be nice.

At the same time, I'm sure people would prefer not seeing a restock bill in the tens of millions when the outfitter prompts to restock their nuclear arsenal :grin:

**Comment by Pointedstick** [comment]
[Showerthoughts incoming]

Small launchers would work if the ammo storage outfits took up weapon space and were increased in size and capacity to prevent the combination of a small launcher + ammo boxes from being better than standalone large launchers. For example two rocket racks are 14 tons and hold the same number of rockets as a 20-ton launcher; unless the two-rocket pylon launcher was 7 tons or more, it would make more sense to buy the smaller launchers plus storage than it would the larger launchers in the first place. But 7 tons for a pylon that only holds two rockets seems like a really lousy deal.

It feels like I'm being led towards re-implementing https://github.com/endless-sky/endless-sky/pull/3283 in a different way. I still feel the small-launchers-large-storage approach makes the most sense for this game.

**Comment by endless-sky** [comment]
I'd prefer not to create a situation where anyone who wants to use a secondary weapon system in anything but a tiny ship has to buy a launcher *plus* ammo *plus* some number of ammo racks. People already get confused enough with the distinction between the ammo and the launcher.

It might be reasonable to have smaller launchers for sale in pirate systems or something as long as a small launcher plus ammo expansions is less space-efficient than the normal launcher. But I wouldn't want a player at a "normal" outfitter to be faced with a confusing decision between two different kinds of launchers for each ammo.

**Comment by Pointedstick** [comment]
Then it seems like we're stuck:

- The status quo of big launchers with large capacity make missile-launching fighters too powerful
- Small launchers with very little capacity might confuse some players due to the necessity of adding ammo storage outfits
- Big launchers with large capacity plus small launchers with little capacity might confuse players about the difference between the launchers

There's always a reason not to change anything, but in this case, I feel like I've made a variety of pretty good cases for how to change missile launchers, and it seems like most folks agree that the status quo has problems of its own.

Also it's hard for me to imagine someone getting confused about the difference between a launcher and its ammo, unless we're talking about young kids here. I remember myself once being a kid playing advanced video games and having no idea how to do a lot of things, culminating in the embarrassment of once getting stuck in _Fallout_ because I never figured out that you could change the cursor by right-clicking. But that wasn't really the game's fault; I just had a tiny, underdeveloped brain--and something like the distinction between a missile launcher and the missiles it fires seems even more elementary.

**Comment by jafdy** [comment]
Giving it a description that says something like `This very small launcher can only fit a single missile before being reloaded making it useless on anything larger than a fighter.` seems like it could work. I always remember putting far to much faith in the descriptions of things so I would see that and then never put it on anything larger than a fighter.

**Comment by Pointedstick** [comment]
Yep, in the branch I'm working on and testing, my small launchers all have text like that.

Alternatively, if we did https://github.com/endless-sky/endless-sky/pull/3283, it would be a simple matter to have every launcher description mention that storage was required to boost capacity beyond what a fighter would carry.

**Comment by Amazinite** [comment]
#3283 seems too drastic a change to me, I think that mini-launchers purpose built for smaller ships is a much better alternative to shrinking launcher sizes and forcing players to install storage units if they want to be in any way serious with their missile use. Even with mini-launchers I'm iffy though, as in their current state, buying mini-launchers would be useless. Perhaps I would be more on board if we went through what MZ said in this comment: https://github.com/endless-sky/endless-sky/issues/3268#issuecomment-346879302 (halving RPS and doubling damage to keep DPS the same but make each individual missile more dangerous). This would make mini-launchers actually something worth considering, as they could have a high power density but low reuse value before needing to reload and still end up being potent weapons, at least for the beginning of a battle. Heck, maybe we should even quadruple damage and fourth RPS, then a fighter carrying only 6 missiles would really be something to worry about. (If we did that, I can imagine we would also make these mini-launchers have the same damage as their larger counterparts but faster fire rates in order for them to just wreck havoc early on.)

**Comment by Pointedstick** [comment]
I largely agree with everything you've said, @Amazinite. I'm going to continue making mini-launchers and I'll have a PR ready in a couple of days, hopefully. Same damage but faster fire rates is actually something I'm already doing for them, so obviously it's a downright genius idea :nerd_face: 

All of this will become really interesting once fighters can re-arm on their carriers, as this will actually make fighters better at something than hyperdrive-capable interceptors with the exact same weapon loadout.

**Comment by Amazinite** [comment]
How are you going to deal with the situation of non-fighters using mini-launchers with higher fire rates and therefore higher DPS but without the ammo limitation given the fact that they can use storage outfits?

**Comment by Pointedstick** [comment]
To be honest, I haven't yet come up with a bulletproof solution to that issue. Higher reload speed might not be necessary anyway if we quadruple damage and quarter the reload times. Just picture it: fighters would streak towards their target, loose a volley of frighteningly powerful missiles, follow up with guns while the launchers reload, then turn around and do it again until they fly back to the mothership to reload once empty. Not only is that more realistic, it sounds really, really fun and awesome.

**Comment by Pointedstick** [comment]
But regardless of how we implement that, I still feel like https://github.com/endless-sky/endless-sky/pull/3283 or a tweaked version is the right path forward:

- It doesn't require any new outfits
- It re-uses an existing mechanic (ammo storage) rather than adding a new one (multiple launchers for one type of ammo)
- It actually does meaningfully limit fighters' capacity; without increasing the storage boxes sizes, mini-launchers don't actually solve the original problem of excessive ammo when armed with missiles--and may actually make it worse since storage boxes use outfit space rather than weapon space, which is more precious

**Comment by Pointedstick** [comment]
I keep trying but I just can't make separate mini-launchers work without increasing ammo boxes' sizes to prevent excessive fighter missile capacity. And if I do that, I've essentially re-done https://github.com/endless-sky/endless-sky/pull/3283 with more complication. It's not a win.

Hardpoint-mounted individual missiles still seems like it could work if missiles were better. But right now, it's just not worth it, even with heavy rockets and lots of extra gun hardpoints.

**Comment by Lineth** [comment]
#3283 honestly does sound like the best way to go about this. The mini-launchers is just the same thing but more convoluted and confusing. Especially if missiles double to quadruple in damage and reload times (which I heavily think they should, plus the hit force buff).

**Comment by Amazinite** [comment]
@Pointedstick I've thought about giving launchers a capacity-like attribute that allows the installation of storage outfits, which means that no launcher means no storage outfits, and mini-launchers would not have this attribute and therefore would be unable to increase their capacities. The only problem with this is that it forces a carrier to install the same launcher that its fighters are using if you want to use the carrier to reload the fighters.

**Comment by Pointedstick** [comment]
@Amazinite I brought that up above. That approach requires code changes, introduces a potentially confusing new game mechanic, and requires the addition of new launchers with new art assets. https://github.com/endless-sky/endless-sky/pull/3283 is overall a much simpler way to implement this.

**Comment by Tadrix** [comment]
> The only problem with this is that it forces a carrier to install the same launcher that its fighters are using if you want to use the carrier to reload the fighters.

Why fighters, while docked, couldn't bypass this limitation? It's not like they are fighting at the time. But reload could take some time.

Besides, modern fighters tend to carry not 2–4, but pretty decent number of missiles ([an example](https://en.wikipedia.org/wiki/General_Dynamics_F-16_Fighting_Falcon#Specifications_(F-16C_Block_50))), and tend to have wide selection of them.

For game purposes I'm thinking about two hardpoints with 10–16 Sidewinder missiles total (5–8 per pod or rack; other missile types vary accordingly), 6 Korath mines (3 per mine launcher), or a single payload bay (taking around 12 weapon space) with 10 Thunderheads, 8 heavy missiles, 4 torpedoes, 2 Typhoons or 2 Finishers per fighter.

**Comment by Pointedstick** [comment]
> For game purposes I'm thinking about two hardpoints with 10–16 Sidewinder missiles total (5–8 per pod or rack; other missile types vary accordingly)

That's _exactly_ what I implemented in https://github.com/endless-sky/endless-sky/pull/3283 :slightly_smiling_face: . Sidewinder capacity becomes [5 per launcher](https://github.com/endless-sky/endless-sky/pull/3283/files#diff-79d98f0b5ecf543ca7b61f2673c2cb50R677); two launchers means 10 missiles. Launchers for bigger missiles hold fewer; launchers for smaller missiles hold more.

**Comment by Tadrix** [comment]
By the way, there could be a special missile secondary weapon for fighters. Takes one gun slot, contains several missiles, not reusable (you can't even buy its ammo separately), expendable. Meaning, when last missile is spent, the outfit is gone (like the nuke in this regard), its casing jettisoned. To replenish, buy a new one from outfitter.

**Comment by Pointedstick** [comment]
That would require code changes, introduce a new game mechanic that could confuse people, and not really offer anything (IMHO) over the fairly simple approach outlined in https://github.com/endless-sky/endless-sky/pull/3283.

**Comment by Pointedstick** [comment]
So right now we have most fighters mounting a single missile launcher--generally a Meteor launcher, but a few have Sidewinders. It works pretty well. The only thing I think is still applicable here is to somehow allow small fighters to additionally carry a small payload of larger missiles without having to mount a huge launcher. Like a rack of four heavy rockets or torpedoes that takes up 5 tons of weapon space but is unable to increase its ammunition capacity from the ammo containers. This would work well if the PR to increase fighter gun ports (https://github.com/endless-sky/endless-sky/pull/3350) gets merged.

It would probably also require some AI improvements to so that fighters could be a bit more discriminating regarding which targets they choose for their missiles, or else you'd see them firing torpedoes at other fighters.

**Comment by TheMarksman-ES** [comment]
>Like a rack of four heavy rockets or torpedoes that takes up 5 tons of weapon space but is unable to increase its ammunition capacity from the ammo containers. This would work well if the PR to increase fighter gun ports (#3350) gets merged. 

Instead of just using existing racks of rockets and/or torpedoes, it might also be good to add new types of anti-capital ship to bolster fighter weaponry.

**Comment by Amazinite** [comment]
Ref: #4392 give certain fighters built-in launchers? Such launchers could be unique outfits, i.e. designed for that fighter, meaning we'd avoid the [previously mentioned](https://github.com/endless-sky/endless-sky/issues/3372#issuecomment-350411806) potential issue of installing high DPS fighter launchers on larger ships. Or does that do too much to remove customization?

**Comment by Zitchas** [comment]
Is there any real problem with having both small launchers and large launchers that use the same ammo? I really don't think that will be confusing. (As far as that goes, I rather liked the "small launcher + racks" method suggested by PointedStick. I feel confident that most people should be able to figure out that concept.

**Comment by Amazinite** [comment]
The idea was to make some distinction between small launchers and large launchers other than just capacity and size. Otherwise small launchers would either be entirely useless, or fighters would need to be changed to accommodate them, likely be increasing their gun ports (ref: #3264).

**Comment by Zitchas** [comment]
Even if they are only intended for fighters, I could see people employing them on interceptors or for limited use applications. For instance, not everyone wants to be able to do a barrage of EMPs (I find them more of a hazard than a help once melee starts), but having one or two used while closing could be very useful. But taking up all that space for the launcher for just a couple missiles is a waste. I would happily sacrifice 75% of the capacity of most launchers in exchange for a 65% reduction in outfit space.

**Comment by Pointedstick** [comment]
I still think my original idea of making all the launchers smaller with reduced capacity was the best. It was simple, allowed for the idea here in an elegant way, didn't require any new outfits. The only real objection made to this was that it could be confusing for players to have to figure out that they need to buy additional outfits to increase their missile capacity. But I think it would be equally confusing to teach players about fighters with built-in weapons, the differences between large and small launchers (particularly if the small launchers are somehow made unable to benefit from the capacity boosting outfits) or hardpoint-mounted missiles (other than nukes of course, but those are much later in the game from the point where a new player can get confused). The missile capacity outfits are pretty common and obvious, I think--and could be made even more obvious with clearer text.

Basically I would just bring back https://github.com/endless-sky/endless-sky/pull/3283 and also edit the launcher and ammo box descriptions to make it even clearer how to get more missile capacity, as I originally did in https://github.com/endless-sky/endless-sky/pull/3275/.

**Comment by Zitchas** [comment]
I'm completely in support of that. I think the idea of "launcher launches things, clips/containers/racks hold stuff" is pretty straight forward - and already exists in-game. It is just a matter of adjusting values.

**Comment by Pointedstick** [comment]
Yeah. I don't have much time for Endless Sky anymore though, sadly. Feel free to take over if you want to push the idea forward.

**Comment by Brick63** [comment]
I actually made a smaller launcer intended for the navy, specifically combat drones and lances, and i basically only need stats for it, i think @Amazinite had something for it at one point
![barrage icon@2x](https://user-images.githubusercontent.com/32852156/75613306-0c207900-5afa-11ea-8244-2d507c6be31c.png)
![barrage launcher@2x](https://user-images.githubusercontent.com/32852156/75613307-0cb90f80-5afa-11ea-852c-7d5ffcef7bd5.png)
![barrage sprite@2x](https://user-images.githubusercontent.com/32852156/75613308-0cb90f80-5afa-11ea-88f8-eb3df01a41cc.png)
![barrage storage@2x](https://user-images.githubusercontent.com/32852156/75613309-0d51a600-5afa-11ea-8a3f-65a8d84619e6.png)

**Comment by Pointedstick** [comment]
That second one looks like some kind of "Missile factory" or "Drone swarm factory" outfit!

**Comment by Brick63** [comment]
It's the storage outfit for the Barrage launcher

**Comment by Maddremor** [comment]
Would it be practical to add a very slow fire rate (if one small launcher has a single missile) or a small magazine with a very long reload time (for outfits with several missiles launched in rapid succession)? This would simulate the downsides with a simple pylon, compared to larger launchers with auto-loaders, as reloading essentially entails attaching a new missile on the hull. This is something that a fighter pilot is unlikely to manage in timely fashion, but it's still useful for an alpha strike with a fighter, or a vessel that is strapped for weapon space, but has some mounting points to spare.

**Comment by Pointedstick** [comment]
Implemented with https://github.com/endless-sky/endless-sky/pull/5793.
